### PR TITLE
Add suspicious group

### LINKED ISSUE #6366: Missing lint categories?
With my recent return to reviewing clippy PRs, I found that it is often hard to categorize lints.

For example, #6086 would warn about something that is *likely* incorrect, but doesn't warrant a `correctness` lint, because it is just a heuristic and could have false positives. For now, we chose `style`, but that seems wrong. Perhaps `consistency`?

### PR DESCRIPTION
**Pull request by camsteffen**
changelog: Introduce `clippy::suspicious` 🤔 group and move several lints into the group

Closes #6366. CC #6626.

A number of lints are moved from each of `correctness`, `style` and `complexity` groups. Notably I didn't move `suspicious_splitn` since I think that is a `correctness` lint despite the name.

Lints moved to `clippy::suspicious`:
* `blanket_clippy_restriction_lints` (was `clippy::style`)
* `empty_loop` (was `clippy::style`)
* `eval_order_dependence` (was `clippy::complexity`)
* `float_equality_without_abs` (was `clippy::correctness`)
* `for_loops_over_fallibles` (was `clippy::correctness`)
* `misrefactored_assign_op` (was `clippy::complexity`)
* `mut_range_bound` (was `clippy::complexity`)
* `mutable_key_type` (was `clippy::correctness`)
* `suspicious_arithmetic_impl` (was `clippy::correctness`)
* `suspicious_assignment_formatting` (was `clippy::style`)
* `suspicious_else_formatting` (was `clippy::style`)
* `suspicious_map` (was `clippy::complexity`)
* `suspicious_op_assign_impl` (was `clippy::correctness`)
* `suspicious_unary_op_formatting` (was `clippy::style`)

### CODE REVIEW

**Comment by flip1995** [comment]
Let's talk about this in our meeting later today, before we finalize this.

**Review by xFrednet — COMMENTED** [review]
The changes look good to me. I focused on the addition in the metadata collection monster and that looks correct. It's nice to see that it's as simple as planned :)

**Comment by flip1995** [comment]
In [today's meeting](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/Meeting.202021-06-15/near/242747922) there was no veto to going forward with this. As Manish pointed out, I'll go through the [Clippy RFC](http://rust-lang.github.io/rfcs/2476-clippy-uno.html) to see if something prevents us to do this, but we can probably do this.

File: README.md [hunk 1]
```diff
@@ -17,6 +17,7 @@ You can choose how much Clippy is supposed to ~~annoy~~ help you by changing the
 | `clippy::style`       | code that should be written in a more idiomatic way                     | **warn**      |
 | `clippy::complexity`  | code that does something simple but in a complex way                    | **warn**      |
 | `clippy::perf`        | code that can be written to run faster                                  | **warn**      |
+| `clippy::suspicious`  | code that seems very likely to be incorrect                             | **warn**      |
```
Comment (flip1995) [inline README.md:20]: I would put this right under `correctness` and word it similar:

```text
| `clippy::suspicious`  | code that is very likely wrong or useless                        | **warn**      |
```

**Review by flip1995 — COMMENTED** [review]
Reading the RFC, there's nothing preventing us from adding lint groups. Though, we can't really remove a group after adding it.

I still have to look at the lints separately.

File: README.md (hunk 1 above)
Comment (camsteffen) [reply README.md:20]: Good idea. How about this? (removed a "very" and changed to "most likely")

```text
`clippy::correctness` | code that is outright wrong or useless
`clippy::suspicious`  | code that is most likely wrong or useless
```

**Comment by llogiq** [comment]
I like it! Not sure if any lints are missing from the list, but those who are there seem to be about right.

File: README.md [hunk 2]
```diff
@@ -13,7 +13,8 @@ You can choose how much Clippy is supposed to ~~annoy~~ help you by changing the
 | Category              | Description                                                             | Default level |
 | --------------------- | ----------------------------------------------------------------------- | ------------- |
 | `clippy::all`         | all lints that are on by default (correctness, style, complexity, perf) | **warn/deny** |
-| `clippy::correctness` | code that is outright wrong or very useless                             | **deny**      |
+| `clippy::correctness` | code that is outright wrong or useless                                  | **deny**      |
+| `clippy::suspicious`  | code that most likely wrong or useless                                  | **warn**      |
```
Comment (flip1995) [inline README.md:17]:
```suggestion
| `clippy::suspicious`  | code that is most likely wrong or useless                               | **warn**      |
```

**Review by flip1995 — APPROVED** [review]
Sorry for taking so long to review. The lints that were moved LGTM. One typo in the readme is left, otherwise r=me (after squashing the fixups).

**Comment by flip1995** [comment]
Oh, it would also be great, if you could edit the PR description to include from which group a lint was moved for every lint, so that writing the changelog for this will be easier :)

**Comment by camsteffen** [comment]
Added remark-gfm in order to allow the table on the readme to exceed the maximum width. See https://github.com/remarkjs/remark-lint/issues/249#issuecomment-708528768. Needed this after updating the description of clippy::all to mention suspicious.

**Comment by camsteffen** [comment]
> Oh, it would also be great, if you could edit the PR description to include from which group a lint was moved for each lint, so that writing the changelog for this will be easier :)

Done! I'll let you review one more time for the remark change.

**Comment by flip1995** [comment]
Remark changes also LGTM.

Thanks! @bors r+

**Pull request merged**